gui-tk/godley.tcl

```diff
@@ -96,6 +96,11 @@ proc openGodley {id} {
         .$id.menubar.edit add command -label Cut -command "$id.cut" -accelerator $meta_menu-X
         .$id.menubar.edit add command -label Copy -command "$id.copy" -accelerator $meta_menu-C
         .$id.menubar.edit add command -label Paste -command "$id.paste" -accelerator $meta_menu-V
+	     if {![catch {clipboard get -displayof .$id.context -type UTF8_STRING}]} {
```

Comment:
The problem with this approach is that your setting the menu item state according to the state of the clipboard at window creation. What you need to do is set the state at menu post time.

Reply:
I've been struggling with this one most of today. The online [documentation](https://www.tcl.tk/man/tcl8.4/TkCmd/menu.htm#M7)  is not clear. The furthest I've got is:

In godley.tcl:90

`        menu .$id.menubar.edit -postcommand {
	     if {[catch {clipboard get -displayof .$id.context -type UTF8_STRING}]} {
			.$id.menubar.edit entryconfigure end -state normal
			}
		}             
        .$id configure -menu .$id.menubar  
        .$id.menubar.edit add command -label Undo -command "$id.undo 1" -accelerator $meta_menu-Z
        .$id.menubar.edit add command -label Redo -command "$id.undo -1" -accelerator $meta_menu-Y
        .$id.menubar.edit add command -label Title -command "textEntryPopup .godleyTitle {[$id.godleyIcon.table.title]} {setGodleyTitleOK $id}"
        .$id.menubar.edit add command -label Cut -command "$id.cut" -accelerator $meta_menu-X
        .$id.menubar.edit add command -label Copy -command "$id.copy" -accelerator $meta_menu-C
        .$id.menubar.edit post 5     
        .$id.menubar.edit add command -label Paste -command "$id.paste" -accelerator $meta_menu-V`

which gives me the following error when I click on Edit in the Godley window menubar:

can't read "id": no such variable
    while executing
".$id.menubar.edit entryconfigure end -state normal"

Reply:
This is one of the most irritating things about TCL - understanding how expansion works, ie the interaction between $, {}, "" and [].

In this case, $id needs to be evaluated at menu instantiation time, not at menu post time. Wrapping a $id in a {} defers expansion until  the code with the braces is run, ie at post time.

The best bet is to create another proc, say editProc, eg:
proc editProc id {
   if {[catch ...
}

and then set menu .$id.menubar.edit -postcommand "editProc $id". Note the use of quotes, which causes id to be immediate expanded, ie at menu instantiation time.

Reply:
Thanks, this has helped a lot. I've also now seen the -postcommand example applied to getBookMarkMenu in minsky.tcl, which is where I decided to start since it is easier than the godley.tcl case. So, now I don't get any error messages, but the last entry on the cascade menu gets greyed out and not the one I want. Here's what I did and thought would work for me:

menu .menubar.edit -postcommand "editProc window"
.menubar add cascade -menu .menubar.edit -label Edit -underline 0

.menubar.edit add command -label "Undo" -command "undo 1" -accelerator $meta_menu-Z
.menubar.edit add command -label "Redo" -command "undo -1" -accelerator $meta_menu-Y
.menubar.edit add command -label "Cut" -command cut -accelerator $meta_menu-X
.menubar.edit add command -label "Copy" -command "minsky.copy" -accelerator $meta_menu-C
.menubar.edit add command -label "Paste" -command "minsky.paste" -accelerator $meta_menu-V
.menubar.edit add command -label "Group selection" -command "minsky.createGroup" -accelerator $meta_menu-G
.menubar.edit add command -label "Dimensions" -command dimensionsDialog

proc editProc {window} {
	if {[catch {clipboard get -displayof $window -type UTF8_STRING}]} {
	.menubar.edit entryconfigure 4 -state disabled
	}
}

I thought that placing the index 4 (counting from 0) after entryconfigure would grey out the corresponding entry (Paste) of the edit cascade menu? I think I'm finally on the right track and hope to have it working next time you take a look at it.

Reply:
I got this functionality working in godley.tcl. Still having the same problem in the minsky.tcl I mentioned in the previous post. Working on it.